prebid-server/pbs-modules/rules-engine.md

```diff
@@ -0,0 +1,805 @@
+
+1. The first level is 'Dimension 1'. It has 3 specific values and one wildcard "other" value.
+2. The second level is 'Dimension 2'. It has 2 or 3 values for each scenario and possibly a wildcard value.
+3. There are 12 specific leaf nodes and one overall default leaf.
+4. Note that the second branch does not have a wildcard option at the second level, so if none of the 3 specific values for Dimension 2 match (red lines), the module will jump to the default leaf.
+
+{: .alert.alert-info :}
+Tip: you can also think of the rule tree as an N-dimensional matrix where each intersection of schema function values resolves to a 'results' node.
+
+See the [runtime processing](#runtime-processing) section below for more details.
+
+### Schema Functions
+
+The module provides a set of functions that parse the OpenRTB request for specific fields, extracting a particular value. For instance, finding out which specific country or channel is associated with this request. These functions are called 'schema functions' because they form the structure of the Rule Tree as described in the section above.
+
+All of these functions operate only on the OpenRTB request at the 'Processed Auction Request' stage of module invocation.
+
+{: .table .table-bordered .table-striped }
+| Schema Function | Args | Description | Output Example |
+| :---- | :---- | :---- | :---- |
+| deviceCountry | null  | Returns OpenRTB device.geo.country. | "CAN" |
+| deviceCountryIn | array of strings representing country codes | If the parameter is an array of strings, then check to see if device.geo.country is part of the named set (case sensitive). If yes, then return "true", else return "false". The parameter must be an array of strings. No length limit. If the array is length zero, treat it as 'null' above. | "true" |
+| datacenters | null  This is an option for those who don't have geo-lookup available on every request. | Returns the value of the datacenter, which is host-company specific. Please work with your host company to get the list of values available in your environment. | "eu-west" |
+| datacentersIn | array of strings containing PBS host company-specific code for region.  This is an option for those who don't have geo-lookup available on every request. | Arguments must be an array of strings with no length limit. The function returns "true"/"false" if the actual datacenter is on the array (case sensitive). Please work with your host company to get the list values available in your environment. | "true" |
+| channel | null | Returns ext.prebid.channel – e.g. web, amp, app. Returns empty string if there is no channel. | "app" |
+| eidAvailable | null  | Returns "true" if user.eids array exists and is non-empty. Returns "false" if user.eids doesn't exist or is empty. | "false" |
+| eidIn | array of strings meant to be extended ID source values like "pubcid.org" | Returns "true"/"false" if the named eid.source is on the array (case sensitive). | "true" |
```

Comment:
Same comment as `deviceCountryIn` regarding array of strings with at least one element